heirloom/expand/unexpand.1
.\"
.\" Copyright (c) 2003 Gunnar Ritter
.\"
.\" This software is provided 'as-is', without any express or implied
.\" warranty. In no event will the authors be held liable for any damages
.\" arising from the use of this software.
.\"
.\" Permission is granted to anyone to use this software for any purpose,
.\" including commercial applications, and to alter it and redistribute
.\" it freely, subject to the following restrictions:
.\"
.\" 1. The origin of this software must not be misrepresented; you must not
.\"    claim that you wrote the original software. If you use this software
.\"    in a product, an acknowledgment in the product documentation would be
.\"    appreciated but is not required.
.\"
.\" 2. Altered source versions must be plainly marked as such, and must not be
.\"    misrepresented as being the original software.
.\"
.\" 3. This notice may not be removed or altered from any source distribution.
.\"
.\" Sccsid @(#)unexpand.1	1.1 (gritter) 5/5/03
.TH UNEXPAND 1 "5/5/03" "Heirloom Toolchest" "User Commands"
.SH NAME
unexpand \- convert spaces to tabs
.SH SYNOPSIS
\fBunexpand\fR [\fB\-a\fR]
[\fB\-t\ \fItablist\fR | \fB\-\fItablist\fR] [\fIfile\fR\ ...]
.SH DESCRIPTION
The
.I unexpand
utility converts multiple space characters found
at the beginning of each line
of its input
.I files
(or of standard input if no
.I file
argument is present)
to tabulators if appropriate
and writes the result to standard output.
The input is ordered in columns
according to the width of characters;
a backspace character causes the column to be decremented.
By default, a tabulator stop is assumed at every eight positions.
.PP
The following options alter the behavior of
.IR unexpand :
.TP
.B \-a
Causes conversion also to apply
to sequences of two or more spaces
found after the first character on each line other than a space.
.TP
.B \fB\-t\ \fItablist\fR
Specify the positions of tabulator stops.
If
.I tablist
is a single number,
a tabulator stop is assumed at every
.I tablist positions.
If
.I tablist
is a list of numbers
in strictly ascending order
separated by commas or blanks,
a tabulator stop is assumed at each numbered position.
For column positions beyond the last tabulator stop,
each tabulator character is converted to a single space character.
Implies
.IR \-a .
.SH "ENVIRONMENT VARIABLES"
.TP
.BR LANG ", " LC_ALL
See
.IR locale (7).
.TP
.B LC_CTYPE
Specifies the mapping of bytes to characters
and the width of characters.
.SH "SEE ALSO"
expand(1),
tabs(1)


ex-vi/Changes
Release ...
* When a multicolumn character was replaced by another multicolumn character
  in insert mode, the display was not updated appropriately with terminals
  other than xterm.
* Various rendering errors happened with multicolumn characters if they
  started at an even column (counting from 1 upwards).
* When a multicolumn character was inserted and then replaced, the visual
  screen representation was sometimes not updated accordingly.
* Undoing the replacement of a multicolumn character by a singlecolumn
  character twice made the singlecolumn character invisible.
* The 'cw' visual command with a multibyte character as last character of
  the affected word located at the end of the line left garbage bytes past
  the end of the new text.
* Visual 'U' followed by 'u' lead to garbage on the screen when multibyte
  characters were present on the changed line.
* The handling of multicolumn characters was fixed for terminals without the
  IM (insert mode) capability. It is unlikely that such terminals actually
  exist, but vi will use the corresponding code for open mode if a termcap
  entry is not available.
* When an illegal multibyte sequence is entered in vi insert mode, no garbage
  bytes are generated anymore when the insert mode is left.
* The '\u', '\l', '\U', and '\L' substitution sequences work with multibyte
  characters now.
* Handle character case conversions with the '~' vi command correctly if the
  length of the converted multibyte sequence is smaller than the original one.
* Multibyte sequences that correspond to an unprintable character are now
  printed as multiple octal escape sequences.
* Fixed an old vi bug: If a vi command that yanked or deleted part of a line
  was followed by an ex command that also yanked or deleted some text, a
  following 'p' vi command pasted the text affected by the former vi command.
  It now pastes the text of the last yank or delete even if that was an ex
  command.

Release 12/2/04
* Support for multibyte character locales was added.
* The code has been converted to ANSI C, and support for pre-POSIX systems has
  been dropped.
* When the end of the current line consists of blank characters and the 'w'
  visual command is given at this point, vi now advances to the start of the
  next line instead of ringing the bell. This is compatible with SVR4 vi and
  seems to be what POSIX specifies.
* If the replacement part of a substitute command consists of a single '%',
  as in ':s/foo/%/', the replacement part of the previous substitution is
  used. This is compatible with SVR4 vi and is specified by POSIX.
* Fixed a number of possible heap overflows, e.g. because of too long tag
  strings.

Release 6/5/04
* Some changes were made to the Makefile to support RPM builds. In particular,
  the meaning of the DESTDIR and PREFIX variables was changed.
* An insufficient size of a variable caused the window size set to 8 on
  FreeBSD if the terminal baud rate was 38400.

Release 1/3/04
* Changes to the included libterm only: Made multiple tc= capabilities in
  a termcap entry work; recognize tc= also if it is not the last capability
  in an entry (Bugreport by Andrew Minter).

Release 9/3/03
* The code did not check st_blksize from stat(2) at other points besides
  the one fixed in the last release.
* The keyboard input character with code 255 ("y in ISO-8859-1) was
  misinterpreted as end-of-file indicator.

Release 8/27/03
* Compile fixes for AIX and HP-UX (Mike Jetzer).
* Delete temporary file when preserving was successful after receiving
  SIGHUP or SIGTERM (Fix taken from 4.3BSD Reno).
* Set MAILRC to /dev/null in expreserve to avoid reading the user's
  mail initialization file.
* Optionally use Caldera's 'UNIX(R) Regular Expression Library' to
  get POSIX.2 compatible REs.
* Don't refuse to quit with 'No write since last change' if a line of a
  newly read input file is too long but no changes were made by the user.
* The POSIX_2 preprocessor define has been removed. The behavior previously
  enabled by this variable is now the default (except as noted below).
* Backslash inside RE bracket expresssions (as in [\]]) is now recognized
  as an escape character by default. This is not POSIX.2 compliant, but is
  compatible with historic vi behavior, is demanded by the some traditional
  standards such as the System V Interface definition, and is compatible
  with vim. To get POSIX.2 compliant behavior here, define NO_BE_BACKSLASH.
* The input buffer did overflow with large values of st_blksize from stat(2).

Release 4/3/02
* Option "showmode" has no effect on hardcopy or glass terminals (P).
* Fixed undo for :r! command.
* Interrupt character is not misinterpreted as DEL on insertion (P).
* Fixed interaction of <count>~ vi command with abbreviations and macros (P).
* Avoid horizontal scrolling of showmode string on wraparound to last line (P).
* No showmode string is printed when executing a macro.
* Recovery listing fixed not to mix characters from long filenames in
  /var/preserve with /var/tmp listing (P).
* Catch SIGXFSZ (file size limit exceeded).

Release 3/3/02
* Separated terse/noterse messages for RE errors (P).
* The expreserve and exrecover commands work again (P).
* Passing beyond EOL in <count>~ vi command is not an error (P).
* Fixed segmentation violation on mdjc'd and related bugs (Sven Mascheck).
* Marks remain on lines changed in an undo operation.
* Close mapmalloc file descriptor for /dev/zero on exec() (P).
* Added -L and -V command line options as in SVr4 ex.
* POSIX.2: Processing EXINIT does not depend on stdin being a terminal.
* POSIX.2: No newline is appended to an empty file in visual mode.
* Fixed segmentation violation on :f^V^M^M and similar commands.
* Mapmalloc extended to allocate and release multiple memory pools.
* Exrecover lists /var/tmp in addition to /var/preserve.
* Have multiple attempts to create named buffer temporary file.
* Size limit for temporary file names removed.

Release 2/17/02

heirloom/expand/unexpand.c
/*
 * unexpand - convert spaces to tabs
 *
 * Gunnar Ritter, Freiburg i. Br., Germany, May 2003.
 */
/*
 * Copyright (c) 2003 Gunnar Ritter
 *
 * This software is provided 'as-is', without any express or implied
 * warranty. In no event will the authors be held liable for any damages
 * arising from the use of this software.
 *
 * Permission is granted to anyone to use this software for any purpose,
 * including commercial applications, and to alter it and redistribute
 * it freely, subject to the following restrictions:
 *
 * 1. The origin of this software must not be misrepresented; you must not
 *    claim that you wrote the original software. If you use this software
 *    in a product, an acknowledgment in the product documentation would be
 *    appreciated but is not required.
 *
 * 2. Altered source versions must be plainly marked as such, and must not be
 *    misrepresented as being the original software.
 *
 * 3. This notice may not be removed or altered from any source distribution.
 */

#if __GNUC__ >= 3 && __GNUC_MINOR__ >= 4
#define	USED	__attribute__ ((used))
#elif defined __GNUC__
#define	USED	__attribute__ ((unused))
#else
#define	USED
#endif
static const char sccsid[] USED = "@(#)unexpand.sl	1.6 (gritter) 12/5/04";

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <libgen.h>
#include <locale.h>
#include <wchar.h>
#include <unistd.h>
#include <ctype.h>

#if defined (__GLIBC__)
#if defined (_IO_putc_unlocked)
#undef	putchar
#define	putchar(c)	_IO_putc_unlocked(c, stdout)
#endif
#endif

#include <iblok.h>
#include "tablist.h"

static int		aflag;
static int		status;
static int		mb_cur_max;
static const char	*progname;

static void	usage(void);
static void	unexpand(const char *);
static void	spaces(long long *, long long *, struct tab **);

int
main(int argc, char **argv)
{
	int	i;

	progname = basename(argv[0]);
	setlocale(LC_CTYPE, "");
	mb_cur_max = MB_CUR_MAX;
	for (i = 1; i < argc && argv[i][0] == '-' && argv[i][1]; i++) {
		if (argv[i][1] == '-' && argv[i][2] == '\0') {
			i++;
			break;
		}
	optc:	if (argv[i][1] == 't') {
			aflag = 1;
			if (argv[i][2])
				settab(&argv[i][2]);
			else
				settab(argv[++i]);
		} else if (argv[i][1] == 'a') {
			aflag = 1;
			if (argv[i][2]) {
				argv[i]++;
				goto optc;
			}
		} else if (isdigit(argv[i][2]))
			settab(&argv[i][1]);
		else
			usage();
	}
	if (tablist)
		aflag = 0;
	if (i < argc)
		while (i < argc)
			unexpand(argv[i++]);
	else
		unexpand(NULL);
	return status;
}

static void
usage(void)
{
	fprintf(stderr, "usage: %s [ -a ] file ...\n",
			progname);
	exit(2);
}

static void
unexpand(const char *fn)
{
	struct iblok	*ip;
	struct tab	*tp;
	char	*cp;
	char	b;
	wint_t	c;
	int	n, w;
	long long	col, spc = 0;

	if ((ip = fn ? ib_open(fn, 0) : ib_alloc(0, 0)) == NULL) {
		perror(fn ? fn : "stdin");
		status |= 1;
		return;
	}
	col = 0;
	tp = tablist;
	while ((mb_cur_max > 1 ? cp = ib_getw(ip, &c, &n) :
				(b = c = ib_get(ip), n = 1,
				 cp = c == (wint_t)EOF ? 0 : &b)),
			cp != NULL) {
		if (c == ' ' && (col == 0 || aflag || tablist && tp)) {
			spc++;
			continue;
		}
		if (spc && (col == 0 || aflag || tablist && tp))
			spaces(&spc, &col, &tp);
		switch (c) {
		case '\n':
			col = 0;
			tp = tablist;
			break;
		case '\b':
			if (col > 0)
				col--;
			break;
		default:
			if (mb_cur_max > 1) {
				if ((w = wcwidth(c)) >= 0)
					col += w;
				else
					col++;
			} else
				col++;
			break;
		case '\t':
			if (tp)
				while (tp && col > tp->t_pos)
					tp = tp->t_nxt;
			if (tp && col == tp->t_pos) {
				col++;
				tp = tp->t_nxt;
			}
			if (tp) {
				while (col < tp->t_pos)
					col++;
				tp = tp->t_nxt;
			} else if (tablist == 0) {
				if (col % tabstop == 0)
					col++;
				while (col % tabstop)
					col++;
			} else
				col++;
			break;
		}
		while (n--) {
			putchar(*cp & 0377);
			cp++;
		}
	}
	if (spc && (col == 0 || aflag || tablist && tp))
		spaces(&spc, &col, &tp);
	if (ip->ib_fd != 0)
		ib_close(ip);
	else
		ib_free(ip);
}

static void
spaces(long long *sp, long long *cp, struct tab **tp)
{
	int	bol;
	long long	nsp;

	bol = *cp == 0;
	if (tablist) {
		while (*tp && (*tp)->t_pos <= *cp)
			*tp = (*tp)->t_nxt;
		while (*sp && *tp && (*cp + *sp) >= (*tp)->t_pos) {
			if (*sp == 1 && !bol)
				putchar(' ');
			else
				putchar('\t');
			*sp -= (*tp)->t_pos - *cp;
			*cp = (*tp)->t_pos;
			*tp = (*tp)->t_nxt;
		}
	} else {
		while (*sp >= (nsp = tabstop - *cp % tabstop)) {
			if (*sp == 1 && !bol)
				putchar(' ');
			else
				putchar('\t');
			*sp -= nsp;
			*cp += nsp;
		}
	}
	while (*sp) {
		putchar(' ');
		(*cp)++;
		(*sp)--;
	}
}
